### ISSUE TITLE
[Package]: Bun

### ISSUE THREAD

**Issue by leap0x7b** [termux/termux-packages#11188]
### Package description

Bun is a new JavaScript runtime with a native bundler, transpiler, task runner and npm client built-in.

### Home page URL

https://bun.sh

### Source code URL

https://github.com/oven-sh/bun

### Packaging policy acknowledgement

- [X] I have read and understand the [Packaging Policy](https://github.com/termux/termux-packages/blob/master/CONTRIBUTING.md#packaging-policy).

**Comment by joshcangit** [comment]
Bun doesn't support Linux aarch64 yet.
Should check out oven-sh/bun#75 for any progress on their end.

**Comment by EAGLE1309** [comment]
Bun has released support for aarch64 so I hope this will get added soon!

**Comment by SunPodder** [comment]
Any update?

**Comment by HaccerCat** [comment]
you should use grun (glibc for termux) to run bun aarch64, use grun ./bun ...

**Comment by holiq** [comment]
> you should use grun (glibc for termux) to run bun aarch64, use grun ./bun ...

How to get the grun(glibc), i have try to `pkg install glibc` and it not found

**Comment by 2096779623** [comment]
> How to get the grun(glibc), i have try to `pkg install glibc` and it not found

https://wiki.termux.com/wiki/Switching_package_manager

https://github.com/termux-pacman/glibc-packages

**Comment by Maxython** [comment]
> > How to get the grun(glibc), i have try to `pkg install glibc` and it not found
> 
> https://wiki.termux.com/wiki/Switching_package_manager
> 
> https://github.com/termux-pacman/glibc-packages

If you only want to use `pacman` for the `gpkg-dev` repo, then it is not necessary to completely switch to `pacman` as the default package manager.

**Comment by holiq** [comment]
It's work, thanks @2096779623 @Maxython

**Comment by joshcangit** [comment]
Maybe `bun` should be added to the **TUR** repository.

**Comment by SunPodder** [comment]
Bun cannot be added to termux yet. I don’t know what’s that grun thing is. But I tried building bun and it requires latest zig which has some breaking changes from what we have in termux. Ig we need to update zig first.

This is the error
```
./build.zig:51:9: error: no member named 'source' in struct 'std.build.Pkg'
[460](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:461)
        .source = pkgPath("src/bun_redirect.zig"),
[461](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:462)
        ^
[462](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:463)
./build.zig:323:32: note: called from here
[463](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:464)
        try addInternalPackages(
[464](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:465)
                               ^
[465](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:466)
./build.zig:207:43: note: called from here
[466](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:467)
pub fn build(b: *std.build.Builder) !void {
[467](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:468)
                                          ^
[468](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:469)
./build.zig:178:79: error: no member named 'mode' in struct 'std.fs.file.OpenFlags'
[469](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:470)
    var runtime_out_file = try std.fs.cwd().openFile("src/runtime.out.js", .{ .mode = .read_only });
[470](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:471)
                                                                              ^
[471](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:472)
./build.zig:12:9: error: no member named 'addIncludePath' in struct 'std.build.LibExeObjStep'
[472](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:473)
    step.addIncludePath("src/deps");
[473](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:474)
        ^
[474](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:475)
./build.zig:320:20: note: called from here
[475](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:476)
        addPicoHTTP(obj, false);
[476](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:477)
                   ^
[477](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:478)
./build.zig:207:43: note: called from here
[478](https://github.com/SunPodder/termux-packages/actions/runs/4214755107/jobs/7315425240#step:6:479)
pub fn build(b: *std.build.Builder) !void {
```

https://devlog.hexops.com/2023/zig-0-11-breaking-build-changes/

**Comment by nacho00112** [comment]
Any updates? I need the package too

**Comment by holiq** [comment]
> Any updates? I need the package too

If see from Bun release, it can install with npm, but idk it work on termux or not

**Comment by joshcangit** [comment]
No, `bun` must be built from source into a package to be able to be installed in Termux.

Using `npm` will error out if used to install `bun` that way.
```
npm ERR! code EBADPLATFORM
npm ERR! notsup Unsupported platform for bun@0.6.9: wanted {"os":"darwin,linux","cpu":"arm64,x64"} (current: {"os":"android","cpu":"arm64"})
npm ERR! notsup Valid os:   darwin,linux
npm ERR! notsup Actual os:  android
npm ERR! notsup Valid cpu:  arm64,x64
npm ERR! notsup Actual cpu: arm64

npm ERR! A complete log of this run can be found in: /data/data/com.termux/files/home/.npm/_logs/2023-06-20T05_55_23_614Z-debug-0.log
```

**Comment by HaccerCat** [comment]
> No, `bun` must be built from source into a package to be able to be installed in Termux.
> 
> Using `npm` will error out if used to install `bun` that way.
> 
> ```
> npm ERR! code EBADPLATFORM
> npm ERR! notsup Unsupported platform for bun@0.6.9: wanted {"os":"darwin,linux","cpu":"arm64,x64"} (current: {"os":"android","cpu":"arm64"})
> npm ERR! notsup Valid os:   darwin,linux
> npm ERR! notsup Actual os:  android
> npm ERR! notsup Valid cpu:  arm64,x64
> npm ERR! notsup Actual cpu: arm64
> 
> npm ERR! A complete log of this run can be found in: /data/data/com.termux/files/home/.npm/_logs/2023-06-20T05_55_23_614Z-debug-0.log
> ```

I just downloaded it from their releases and it work ig.
idk how I got grun, I installed it over a year ago. but it provides glibc runtime iirc
![Screenshot_20230620-090734](https://github.com/termux/termux-packages/assets/86655602/8f207461-1962-4d17-bf87-7d7ae20db5a3)

**Comment by neuthral** [comment]
thanks @Maxython got it working and made an alias 
`alias bun='grun ~/.bun/bin/bun'`

**Comment by n-ce** [comment]
> Bun cannot be added to termux yet. I don’t know what’s that grun thing is. But I tried building bun and it requires latest zig which has some breaking changes from what we have in termux. Ig we need to update zig first.

Latest version of zig is available on termux now.

![20230907_194126](https://github.com/termux/termux-packages/assets/69345507/b1101730-2870-4e76-9545-0eafd7d1827c)

**Comment by SunPodder** [comment]
> > Bun cannot be added to termux yet. I don’t know what’s that grun thing is. But I tried building bun and it requires latest zig which has some breaking changes from what we have in termux. Ig we need to update zig first.
> 
> Latest version of zig is available on termux now. ![zig_Termux](https://user-images.githubusercontent.com/69345507/266345709-e5305c30-97c2-4304-88bb-5a4ff48c35a7.jpg)

Yeah I have looked into it. Currently there's no official way to cross compile bun. Moreover, building bun requires a pre-built bun executable. So there's no way to build it on termux either.

**Comment by SunPodder** [comment]
@HaccerCat @holiq does `grun bun install` work for you? I'm using bun 1.0.0 and it doesn't work.
```
bun install glob
bun add v1.0.0 (822a00c4)

 installed glob@10.3.4 with binaries:
  - glob

Failed to install 41 packages
error: AccessDenied installing bun-types
error: AccessDenied installing glob
error: AccessDenied installing foreground-child
error: AccessDenied installing jackspeak
error: AccessDenied installing minimatch
error: AccessDenied installing minipass
error: AccessDenied installing path-scurry
error: AccessDenied installing cross-spawn
...
```

**Comment by PoneyClairDeLune** [comment]
According to the [official docs](https://bun.sh/docs/project/development)...

> Bun must be installed to compile Bun
> 
> Bun [depends on ESBuild](https://bun.sh/docs/project/development#install-dependencies)

Wow, just wow. Why the heck don't the Bun devs directly use ESBuild to bundle and minify your files instead?

**Comment by Maxython** [comment]
> @HaccerCat @holiq does `grun bun install` work for you? I'm using bun 1.0.0 and it doesn't work.
> 
> ```
> bun install glob
> bun add v1.0.0 (822a00c4)
> 
>  installed glob@10.3.4 with binaries:
>   - glob
> 
> Failed to install 41 packages
> error: AccessDenied installing bun-types
> error: AccessDenied installing glob
> error: AccessDenied installing foreground-child
> error: AccessDenied installing jackspeak
> error: AccessDenied installing minimatch
> error: AccessDenied installing minipass
> error: AccessDenied installing path-scurry
> error: AccessDenied installing cross-spawn
> ...
> ```

Try this:
```
bun install --backend=copyfile glob
```

**Comment by romanovj** [comment]
> @HaccerCat @holiq does `grun bun install` work for you? I'm using bun 1.0.0 and it doesn't work.
> 
> ```
> bun install glob
> bun add v1.0.0 (822a00c4)
> 
>  installed glob@10.3.4 with binaries:
>   - glob
> 
> Failed to install 41 packages
> error: AccessDenied installing bun-types
> error: AccessDenied installing glob
> error: AccessDenied installing foreground-child
> error: AccessDenied installing jackspeak
> error: AccessDenied installing minimatch
> error: AccessDenied installing minipass
> error: AccessDenied installing path-scurry
> error: AccessDenied installing cross-spawn
> ...
> ```

that's because bun tries to create hardlinks

**Comment by HaccerCat** [comment]
> @HaccerCat @holiq does `grun bun install` work for you? I'm using bun 1.0.0 and it doesn't work.
> 
> ```
> bun install glob
> bun add v1.0.0 (822a00c4)
> 
>  installed glob@10.3.4 with binaries:
>   - glob
> 
> Failed to install 41 packages
> error: AccessDenied installing bun-types
> error: AccessDenied installing glob
> error: AccessDenied installing foreground-child
> error: AccessDenied installing jackspeak
> error: AccessDenied installing minimatch
> error: AccessDenied installing minipass
> error: AccessDenied installing path-scurry
> error: AccessDenied installing cross-spawn
> ...
> ```

I did not try 1.0 and actually did not use bun since a long time, but last time I used `bun install` I got a kernel or syscall error I forgot 😆

**Comment by holiq** [comment]
> @HaccerCat @holiq does `grun bun install` work for you? I'm using bun 1.0.0 and it doesn't work.
> 
> ```
> bun install glob
> bun add v1.0.0 (822a00c4)
> 
>  installed glob@10.3.4 with binaries:
>   - glob
> 
> Failed to install 41 packages
> error: AccessDenied installing bun-types
> error: AccessDenied installing glob
> error: AccessDenied installing foreground-child
> error: AccessDenied installing jackspeak
> error: AccessDenied installing minimatch
> error: AccessDenied installing minipass
> error: AccessDenied installing path-scurry
> error: AccessDenied installing cross-spawn
> ...
> ```

No bro, I haven't tried it

**Comment by kaushiksahu18** [comment]
> It's work, thanks @2096779623 @Maxython

Can u please tell me also, what I have to do.
> I am very beginner in linux (in termux too)

** Please provide step by step guide, (commends if possible)**

```
# my state
bun
exec: Failed to execute process '/data/data/com.termux/files/home/.bun/bin/bun': The file exists and is executable. Check the interpreter or linker?
```

**Comment by truboxl** [comment]
None of the distro build and distribute bun
The ones that do just download binaries from upstream and package it that way
https://github.com/oven-sh/bun/issues/862
https://repology.org/project/bun/versions

**Comment by Jobians** [comment]
No updates?

**Comment by twaik** [comment]
No activity usually means no updates.

**Comment by TomJo2000** [comment]
> No updates?

- Nobody has written a build script for `bun` yet.
- The last message is from 7 months ago.
- The package has no existing build scripts in other distros as of 7 months ago.
- And it depends on Kernel version 5.6 (or 5.1 if you're going by absolutely lowest supported version)
So no, no updates.

**Comment by licy183** [comment]
Good news: Bun recently added Android target.

https://github.com/oven-sh/bun/commit/2ee9cad0ea26e051acb181bb3740e292757fdcf5

**Comment by TomJo2000** [comment]
Less good news, someone still needs to write a build script, and the existing examples still consist only of Arch's pkgbuild for bun as all other packages of it that I'm aware of just download the release binaries.

**Comment by Jobians** [comment]
Good news everyone, Bun now works natively on Termux.

Tested and working fine on my device using the Android ARM64 build: https://github.com/oven-sh/bun/issues/8685#issuecomment-4454640901

<img width="1080" height="2412" alt="Image" src="https://github.com/user-attachments/assets/c345647e-de39-4efb-b010-b06567462e12" />

**Comment by Jak2k** [comment]
Bun is completely vibecoded now. I'd recommend switching to deno or plain nodejs.

**Comment by Jobians** [comment]
@licy183 please bro look into it https://github.com/termux/termux-packages/issues/11188#issuecomment-4462380479

**Comment by Jobians** [comment]
@licy183

**Comment by joshcangit** [comment]
> Bun is completely vibecoded now. I'd recommend switching to deno or plain nodejs. 

I'm now looking into trying `nub` or `aube`.

So far, `aube` works but only if downloaded manually from their releases.
As for `nub`, this can't work.
